proxy/proxy.go

```diff
@@ -0,0 +1,173 @@
+package proxy
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	jwt "github.com/dgrijalva/jwt-go"
+)
+
+type Proxy struct {
+	jwtSecret string
+	apis      []*ApiProxy
+}
+
+type ApiProxy struct {
+	name    string
+	matcher *regexp.Regexp
+	handler http.Handler
+	token   string
+	roles   []string
+}
+
+type JWTClaims struct {
+	ID           string                 `json:"id"`
+	Email        string                 `json:"email"`
+	AppMetaData  map[string]interface{} `json:"app_metadata"`
+	UserMetaData map[string]interface{} `json:"user_metadata"`
+	*jwt.StandardClaims
+}
+
+// Error is an error with a message
+type Error struct {
+	Message string `json:"msg"`
+	Code    int    `json:"code"`
+}
+
+var bearerRegexp = regexp.MustCompile(`^(?i)Bearer (\S+$)`)
+
+func New(secret string, apis []*ApiProxy) *Proxy {
+	return &Proxy{
+		jwtSecret: secret,
+		apis:      apis,
+	}
+}
+
+func NewApiProxy(name, urlString, token string, roles []string) (*ApiProxy, error) {
+	matcher := regexp.MustCompile("^/" + name + "/?")
+
+	target, err := url.Parse(urlString)
+	if err != nil {
+		return nil, err
+	}
+	director := buildDirector(target, matcher, token)
+	handler := &httputil.ReverseProxy{Director: director}
+
+	return &ApiProxy{
+		name:    name,
+		matcher: matcher,
+		handler: handler,
+		token:   token,
+		roles:   roles,
+	}, nil
+}
+
+func (p *Proxy) HandleRequest(w http.ResponseWriter, r *http.Request) *Error {
+	for _, api := range p.apis {
+		if api.matcher.MatchString(r.URL.Path) {
+			if err := p.authenticateProxy(w, r, api); err != nil {
+				return err
+			}
+
+			api.handler.ServeHTTP(w, r)
+			return nil
+		}
+	}
+
+	return &Error{"Not Found", 404}
+}
+
+func (p *Proxy) authenticateProxy(w http.ResponseWriter, r *http.Request, proxy *ApiProxy) *Error {
+	if r.Method == http.MethodOptions {
+		return nil
+	}
+
+	authHeader := r.Header.Get("Authorization")
+	if authHeader == "" {
+		return &Error{"This endpoint requires a Bearer token", 401}
+	}
+
+	matches := bearerRegexp.FindStringSubmatch(authHeader)
+	if len(matches) != 2 {
+		return &Error{"This endpoint requires a Bearer token", 401}
+	}
+
+	token, err := jwt.ParseWithClaims(matches[1], &JWTClaims{}, func(token *jwt.Token) (interface{}, error) {
+		if token.Header["alg"] != jwt.SigningMethodHS256.Name {
```

Comment:
it might be worth extracting the `SigningMethodHS256` constant. It isn't clear that this is what is required.